Fix #223: Enhance Network "get_next" methods to optionally allocate/reserve at the same time

rmhasan opened the pull request:
Users can send a post request to `next_network` and `next_address` actions. If it is a post request then the action will get the networks or addresses and then save them as network objects in the database as `allocated`. If the user sends the `reserve` flag and sets it as true, then it will be saved in the `reserved` state.

nickpegg commented on nsot/api/views.py:
```diff
@@ -504,23 +519,41 @@ def next_network(self, request, pk=None, site_pk=None, *args, **kwargs):
+            )
+            self.allocate_networks(networks, prefix_length, site_pk, network, state=state, is_ip=is_ip)
+        else:
+            networks = network.get_next_network(
```
Suggestion: Since you're returning the `unicode` representation of the Networks in all cases, you could simplify this code by getting rid of this `else` and then set `networks` with `as_objects=True` right before the if statement.

nickpegg commented on nsot/api/views.py:
```diff
@@ -504,23 +519,41 @@ def next_network(self, request, pk=None, site_pk=None, *args, **kwargs):
+                    prefix_length = 128
+                self.allocate_networks(addresses, prefix_length, site_pk, network, state=state, is_ip=True)
+        else:
+            addresses = network.get_next_address(num, strict, as_objects=False)
```
Same thing as with `next_network`, it looks like this `else` can be removed if you move the setting of `addresses` to before the if with `as_objects=True`.

nickpegg commented on nsot/api/views.py:
```diff
@@ -504,23 +519,41 @@ def next_network(self, request, pk=None, site_pk=None, *args, **kwargs):
-
-        return self.success(networks)
+        if request.method == 'POST':
+            state = 'reserved' if qpbool(params.get('reserve', False)) else 'allocated'
```
These state strings are set as constants on the Network model, which you should use just in case we ever change the string in the future:
https://github.com/dropbox/nsot/blob/067699cdb28b3122cb36df613a79e208fb37fb9a/nsot/models.py#L690-L693

nickpegg commented on nsot/api/views.py:
```diff
@@ -504,23 +519,41 @@ def next_network(self, request, pk=None, site_pk=None, *args, **kwargs):
-        return self.success(addresses)
+        if request.method == 'POST':
+            addresses = network.get_next_address(num, strict, as_objects=True)
+            state = 'reserved' if qpbool(params.get('reserve', False)) else 'allocated'
```
Same thing as with `next_network`, you should use the Network state constants here

nickpegg commented on tests/model_tests/test_networks.py:
```diff
@@ -376,4 +376,6 @@ def test_strict_allocation_3(site):
     expected = [ipaddress.ip_network(u'2001:db8:abcd:12::8000:0/128')]
-    assert parent.get_next_network(128, strict = True) == expected 
\ No newline at end of file
+    assert parent.get_next_network(128, strict = True) == expected
```
Was this whitespace addition unintentional?

rmhasan replied:
Yeah it was added unintentionally.

rmhasan commented on nsot/api/views.py:
```diff
@@ -424,6 +424,21 @@ class NetworkViewSet(ResourceViewSet):
+                                 site=models.Site.objects.get(pk=site_pk),
+                                 parent=parent,
+                                 state=state,
+                                 is_ip=is_ip)
```
Okay I'll set `editable` for `is_ip` to false.

rmhasan commented:
Hi @nickpegg 
I added the following changes
- I removed setting `is_ip` from the `next_network` and `next_address` views.
- In `model.Network`, for the `is_ip` field I set `editable` to False.
- In the `next_network` and `next_address` views, I use the `RESERVED` and `ALLOCATED` constants from 
  the model.Network class
- In the `next_network` and `next_address` views, I moved the `network.get_next_network` function 
  call out of the if clause, and set the `as_objects` paramter to True. Also I removed the else clause
- Also take a look at dropbox/pynsot#132. It allows users to send post requests to next_network and 
  next_address

jathanism commented on nsot/api/views.py:
```diff
@@ -424,6 +424,23 @@ class NetworkViewSet(ResourceViewSet):
+    def allocate_networks(self, networks, prefix_length, site_pk, parent, state='allocated'):
+        site = models.Site.objects.get(pk=site_pk)
+        for n in networks:
+            obj = models.Network(
```
I think it would be cleaner to remove the `prefix_length` argument from this method and just allow it to take a list of CIDRs.

Because broadcast_address, ip_version, parent are all automatically populated and overridden internally so providing them up front is redundant.

Then this could just be:

```python
for n in networks:
    obj = models.Network.create(cidr=n, site=site, state=state)
```

jathanism commented on nsot/api/views.py:
```diff
@@ -495,32 +512,45 @@ def supernets(self, request, pk=None, site_pk=None, *args, **kwargs):
-        addresses = network.get_next_address(num, strict, as_objects=False)
-
-        return self.success(addresses)
+        addresses = network.get_next_address(num, strict, as_objects=True)
```
If we ditch `prefix_length` as an argument to `.allocate_networks()` we can probably go back to returning a list of unicode strings (`as_objects=False`) and then the version/prefix_length calculation can also go away!

jathanism commented on nsot/api/views.py:
```diff
@@ -495,32 +512,45 @@ def supernets(self, request, pk=None, site_pk=None, *args, **kwargs):
+            else:
+                state = models.Network.ALLOCATED
+            self.allocate_networks(networks, prefix_length, site_pk, network, state=state)
+        return self.success([unicode(x) for x in networks])
```
Same here, too... We could get revert to `as_objects=False` and `return self.success(networks)` here as well as not passing `prefix_length` to `self.allocate_networks(...)`. :)

jathanism commented:
@rmhasan We miss you! <3

rmhasan commented:
Hi @jathanism, sorry I've been gone for so long. I implemented the changes you asked for, please take a look when you get a chance.

jathanism commented:
@rmhasan Awesome! Welcome back. This is great! We'll go ahead and get this released soon. :)
